### ISSUE TITLE
[bug] Slices of small structs don't use shim funcs to encode/decode their fields

### ISSUE THREAD

**Issue by zond** [tinylib/msgp#73]
The shim funcs to encode/decode third party types work beautifully, and to encode/decode slices of structs including these third party types I just have to define explicit slice types.

These slice types usually use the struct type encode/decode funcs to encode/decode their elements, but if the struct type is small enough (I guess the reason to be) the slice type will just inline the field encoding/decoding.

In these cases, the shim methods are _not_ used for the encoding/decoding, and the code won't compile.

**Comment by philhofer** [comment]
Check out #74  and let me know if it fixes things. (I added a test case, but I'm not exactly sure it's what you were describing.) I also took some preliminary measures to address #71.

**Comment by zond** [comment]
Nope.

Here is the `go generate` log:

```
======== MessagePack Code Generator =======
>>> Input: "auth.go"...
parsing AuthTypes...
parsing AuthType...
parsing Auths...
parsing Auth...
parsing Roles...
parsing Role...
applying shim for datastore.Key to String...
inlining methods for datastore.Key into Auth...
inlining methods for datastore.Key into Auth...
inlining methods for datastore.Key into Auth...
inlining methods for datastore.Key into Auth...
inlining methods for datastore.Key into Auth...
inlining methods for Role into Roles...
inlining methods for datastore.Key into Role...
>>> Generating "auth_gen.go"...
>>> Generating "auth_gen_test.go"...
>>> Formatting "auth_gen.go"...
>>> Done.
```

And here is a generated func:

```
// UnmarshalMsg implements msgp.Unmarshaler
func (z *Roles) UnmarshalMsg(bts []byte) (o []byte, err error) {
        var xsz uint32
        xsz, bts, err = msgp.ReadArrayHeaderBytes(bts)
        if err != nil {
                return
        }
        if cap((*z)) >= int(xsz) {
                (*z) = (*z)[:xsz]
        } else {
                (*z) = make(Roles, xsz)
        }
        for wht := range *z {
                var field []byte
                _ = field
                var isz uint32
                isz, bts, err = msgp.ReadMapHeaderBytes(bts)
                if err != nil {
                        return
                }
                for isz > 0 {
                        isz--
                        field, bts, err = msgp.ReadMapKeyZC(bts)
                        if err != nil {
                                return
                        }
                        switch msgp.UnsafeString(field) {
                        case "Id":
                                if msgp.IsNil(bts) {
                                        bts, err = msgp.ReadNilBytes(bts)
                                        if err != nil {
                                                return
                                        }
                                        (*z)[wht].Id = nil
                                } else {
                                        if (*z)[wht].Id == nil {
                                                (*z)[wht].Id = new(datastore.Key)
                                        }
                                        bts, err = (*z)[wht].Id.UnmarshalMsg(bts)
                                        if err != nil {
                                                return
                                        }
                                }
                        case "Name":
                                (*z)[wht].Name, bts, err = msgp.ReadStringBytes(bts)
                                if err != nil {
                                        return
                                }
                        default:
                                bts, err = msgp.Skip(bts)
                                if err != nil {
                                        return
                                }
                        }
                }
        }
        o = bts
        return
}
```

Note how this func (and the `DecodeMsg` func as well) doesn't use the shim for *datastore.Key.

**Comment by zond** [comment]
Note that I applied the concepts from https://github.com/tinylib/msgp/pull/72 before I tried this, but I don't think it made any difference to this case.

**Comment by philhofer** [comment]
Ok, I see what's happening. `Role` gets inlined into `Roles` before `*datastore.Key` gets inlined into `Role`.

**Comment by zond** [comment]
Oh, interesting :D Will it be hard to fix?

And, please take a look at my pull request - I would be happy to modify it to your liking as long as I have the ability to force value receivers :)

**Comment by philhofer** [comment]
Apparently not! One-line fix. Check out the latest commit on #74. 

I'll review #72 asap.

**Comment by zond** [comment]
It worked! Thanks :)

**Comment by philhofer** [comment]
Fixed by #74
